flax/nnx/nn/recurrent.py

```diff
@@ -0,0 +1,920 @@
+
+    using :func:`flax.linen.scan`.
+    """
+
+    def __init__(
+        self,
+        cell: RNNCellBase,
+        time_major: bool = False,
+        return_carry: bool = False,
+        reverse: bool = False,
+        keep_order: bool = False,
+        unroll: int = 1,
+    ):
+        self.cell = cell
+        self.time_major = time_major
+        self.return_carry = return_carry
+        self.reverse = reverse
+        self.keep_order = keep_order
+        self.unroll = unroll
+
+    def __call__(
+        self,
+        inputs: Array,
+        *,
+        initial_carry: Carry | None = None,
+        seq_lengths: Array | None = None,
+        return_carry: bool | None = None,
+        time_major: bool | None = None,
+        reverse: bool | None = None,
+        keep_order: bool | None = None,
+        rngs: rnglib.Rngs | None = None,
+    ):
+        if return_carry is None:
+            return_carry = self.return_carry
+        if time_major is None:
+            time_major = self.time_major
+        if reverse is None:
+            reverse = self.reverse
+        if keep_order is None:
+            keep_order = self.keep_order
+
+        # Infer the number of batch dimensions from the input shape.
+        # Cells like ConvLSTM have additional spatial dimensions.
+        time_axis = 0 if time_major else inputs.ndim - (self.cell.num_feature_axes + 1)
+
+        # make time_axis positive
+        if time_axis < 0:
+            time_axis += inputs.ndim
+
+        if time_major:
+            # we add +1 because we moved the time axis to the front
+            batch_dims = inputs.shape[1 : -self.cell.num_feature_axes]
+        else:
+            batch_dims = inputs.shape[:time_axis]
+
+        # maybe reverse the sequence
+        if reverse:
+            inputs = jax.tree_util.tree_map(
+                lambda x: flip_sequences(
+                    x,
+                    seq_lengths,
+                    num_batch_dims=len(batch_dims),
+                    time_major=time_major,  # type: ignore
+                ),
+                inputs,
+            )
+        if rngs is None:
+            rngs = rnglib.Rngs(0)
+        carry: Carry = (
+            self.cell.initialize_carry(
+                rngs, inputs.shape[:time_axis] + inputs.shape[time_axis + 1 :]
+            )
+            if initial_carry is None
+            else initial_carry
+        )
+
+        slice_carry = seq_lengths is not None and return_carry
+
+        def scan_fn(carry: Carry, x: Array) -> tuple[Carry, Array]:
+            carry, y = self.cell(carry, x)
+            if slice_carry:
+                return carry, (carry, y)
+            return carry, y
+
+        scan = nnx.scan(
+            scan_fn,
+            in_axes=(Carry, time_axis),
+            out_axes=(Carry, (0, time_axis)) if slice_carry else (Carry, time_axis),
+            unroll=self.unroll,
+        )
+
+        scan_output = scan(carry, inputs)
```

Comment:
Currently `self` is being passed as a capture, we need to pass `cell` as an explicit input.
```suggestion
        def scan_fn(cell: RNNCellBase, carry: Carry, x: Array) -> tuple[Carry, Array]:
            carry, y = cell(carry, x)
            if slice_carry:
                return carry, (carry, y)
            return carry, y

        state_axes = nnx.StateAxes({...: Carry})
        scan = nnx.scan(
            scan_fn,
            in_axes=(state_axes, Carry, time_axis),
            out_axes=(Carry, (0, time_axis)) if slice_carry else (Carry, time_axis),
            unroll=self.unroll,
        )

        scan_output = scan(self.cell, carry, inputs)
```